### PR TITLE
Merging Casey's Thesis Work

### PR DESCRIPTION
**Pull request by csdechant**
This PR directly addresses and should close Issue #257. This PR is based on @cticenhour [thesis-rebase](https://github.com/cticenhour/zapdos/tree/thesis-rebase) branch, excluding $H_{\phi}$ objects. The main focus of this PR  includes:
- The includes of a new material object, `FieldSolverMaterial`, which allows uses to supply the electric field terms in the drift-diffusion equation as a electrostatic potential variable or a vector electric field variable.
- A material object to calculate the plasma dielectric coefficient.

This PR also includes updates based on @csdechant [ZapdosAD-plus](https://github.com/csdechant/zapdos/tree/ZapdosAD-plus) (which was rebased early on into the [thesis-rebase](https://github.com/cticenhour/zapdos/tree/thesis-rebase) branch), the excluding FV objects. That work includes:
- Kernel objects for the electron energy flux using a thermal conductivity and an effected electric field calculation for ions.
- Boundary conditions objects using an effected electric field calculation for ions and a dielectric time integral boundary condition.
- Tests based on the method of manufactured solutions (MMS).
- Minor bug fixes to the Newton-Shooting method acceleration objects.

### CODE REVIEW

**Comment by csdechant** [comment]
@cticenhour this PR has a lot of commit to begin with. I am in favor of squashing all the commit into one, but since this PR is based on your old branch, I wanted to get your opinion first.

**Comment by csdechant** [comment]
@cticenhour I am fixing the Precheck errors and one of them is a "banned keywords" error. The `PlasmaDielectricConstant` material object uses `std::cout`. I am suggest that I just comment out the section of code and leave a `NOTE:` comment stating that users can uncomment this section of code for debugging purposes. What are your thoughts?

**Comment by csdechant** [comment]
@cticenhour @gsgall This PR is ready for review. The same as for the other active PR, #263, @gsgall is planned to be the primary reviewer (since this PR involved Casey's and my work). There are two issues I want to point out:
- Depending on which PR is merged first (this one or #263), the other PR will most likely need to be rebased due to replacing `_grad_potential` with `_electric_field`.
- There is a test fail due to a Exodiff for `tutorial/tutorial05-PlasmaWaterInterface`. In particular, the difference is with the `Current_OHm` and `EFieldx1` auxvariables with a relative difference of 3.46855e-05 and 1.67417e-05, respectively. This error only occurs with CIVET and I have not been able to replicated with my Mac (Intel processor). I am in favor of increasing the ceiling threshold for these variables, but I wanted you guys input first.

**Comment by csdechant** [comment]
> * There is a test fail due to a Exodiff for `tutorial/tutorial05-PlasmaWaterInterface`. In particular, the difference is with the `Current_OHm` and `EFieldx1` auxvariables with a relative difference of 3.46855e-05 and 1.67417e-05, respectively. This error only occurs with CIVET and I have not been able to replicated with my Mac (Intel processor). I am in favor of increasing the ceiling threshold for these variables, but I wanted you guys input first.

I was able to replicate some of this test error with my Ubuntu machine (only a difference for `Current_OHm` showed up). The current ceiling for the auxvariable for this test is 9.9e-6. I would recommend to increase it to 4e-5 just for auxvariables.

**Comment by gsgall** [comment]
> @cticenhour @gsgall This PR is ready for review. The same as for the other active PR, #263, @gsgall is planned to be the primary reviewer (since this PR involved Casey's and my work). There are two issues I want to point out:
> 
> * Depending on which PR is merged first (this one or [Updating Doc String: Address Issue #258 #263](https://github.com/shannon-lab/zapdos/pull/263)), the other PR will most likely need to be rebased due to replacing `_grad_potential` with `_electric_field`.
> * There is a test fail due to a Exodiff for `tutorial/tutorial05-PlasmaWaterInterface`. In particular, the difference is with the `Current_OHm` and `EFieldx1` auxvariables with a relative difference of 3.46855e-05 and 1.67417e-05, respectively. This error only occurs with CIVET and I have not been able to replicated with my Mac (Intel processor). I am in favor of increasing the ceiling threshold for these variables, but I wanted you guys input first.

I will have time to work on the reviews for both of these later in the week. I still have a last homework assignment for the semester I am working on finishing.

**Comment by csdechant** [comment]
> That seems fine to me, but It seems like we have to keep raising tolerances on testing. Maybe we should re think how we do testing and move to more robust tests in the future.

That is true and Issue #261 should help with that (currently working on that issue). For more reference with this test (`tutorial/tutorial05-PlasmaWaterInterface`), it is basically a tutorial version of a current test `1d_dc/mean_en`. Looking at the `cmp` files, it seems that `1d_dc/mean_en` uses looser tolerances for the auxvariables that failed in `tutorial/tutorial05-PlasmaWaterInterface` (they are set to 9.6e-5 for `mean_en_out.cmp`). I will set the new tolerances based on `mean_en_out.cmp` and in the future we should try on consolidated these tests.

**Comment by cticenhour** [comment]
@gsgall My two cents - OK with loosening tolerances here in this case (at least to the point we are consistent with a similar test elsewhere in the suite, as @csdechant pointed out), but I am hoping that starting work on re-vamping our V&V tests in #261 will make our approach to testing more granular and more targeted overall. We can't troubleshoot nearly as quickly when most of our tests are large and purely physics-application-driven.

Given the number of platforms we're developing and testing against, I expect some custom CMP files or adjusted tolerances here and there, but we should try not to make loosening tolerances the norm - I agree!

**Comment by cticenhour** [comment]
I don't want to get in the way of @gsgall's review, but I have one concern about the size of the PR. 15MB is way too big. @csdechant - please look into how we can reduce the size of the new gold files by limiting the output we're diffing to only certain times in the simulation.

**Comment by gsgall** [comment]
> I don't want to get in the way of @gsgall's review, but I have one concern about the size of the PR. 15MB is way too big. @csdechant - please look into how we can reduce the size of the new gold files by limiting the output we're diffing to only certain times in the simulation.

I definitely think keeping the gold files size at a minimum should be a priority. If we are adding more capability with this PR then I think a lot of the testing should be really simple V&V testing. Which in my experience only requires small gold files. 

@csdechant Is the testing strategy of this PR similar to the larger physics based testing used for the majority Zapdos testing? Or is it more math based, focusing on the simplest tests possible for each kernel, i.e having a single kernel in an input whenever possible.

**Comment by csdechant** [comment]
> I don't want to get in the way of @gsgall's review, but I have one concern about the size of the PR. 15MB is way too big. @csdechant - please look into how we can reduce the size of the new gold files by limiting the output we're diffing to only certain times in the simulation.

@cticenhour This is a very fair request and I will re-work the gold files. Currently the new tests looks at all time steps, but I will edit it to just look at every $\frac{1}{4}T$ time steps ($T$ being the period of the cycle).

**Comment by gsgall** [comment]
> > @csdechant Is the testing strategy of this PR similar to the larger physics based testing used for the majority Zapdos testing? Or is it more math based, focusing on the simplest tests possible for each kernel, i.e having a single kernel in an input whenever possible.
> 
> @gsgall All new test are math based (MMS to be exact) and in my option, all new tests in Zapdos should be against some type of known solution (either analytical or MMS) and any new test based on validation should be on a case-by-case basis. For the new MMS test, they are designed to test problems of increase coupling (i.e. one for diffusion-only, next with advection-diffusion with a function potential, next with advection-diffusion with variable potential, etc.).
> 
> The reason for the large gold files is that these MMS tests use solutions designed for an RF discharge, so the test are for one RF cycle. The current tests looks at every time step, so the gold files are big (in honesty, I could probably increase the step size too, as the original step size was chosen to be significantly small enough not to interfere with spatial convergence studies).

@csdechant Is there a reason we cannot just a steady state mms test for each Kernel individually? Testing with increasing coupling over an entire RF cycle seems like it is overkill for testing each Kernel and a bit redundant. I don't really see a need to have these tests be a transient simulation if the Kernel is not a time derivative. If we guarantee that each kernel works as expected on its own we should be able to guarantee the expected behaviour of our system, since MOOSE takes care of using multiple kernels in one system. I would like to advocate for a simpler testing approach. A single steady state test MMS test for each kernel where each test uses only as many objects as strictly necessary for testing a specific kernel, ideally a single kernel per input, if possible. I would also like to see our testing convention be something closer to the MOOSE convection of separating the tests by what type of object we are testing (Kernel, AuxKernel, ..., etc.). 

The other concern, along with the size of gold files, I have with these tests is the amount of time it takes to run the test suite. Checking out the CI, testing, [https://civet.inl.gov/job/2565410/](https://civet.inl.gov/job/2565410/), it shows that this test suite takes a little over 7 minutes to run. It also shows an average test time of 16 seconds, even when running the tests in parallel. I really think we should be aiming to decrease this to somewhere around a few seconds for each test, unless there is a specific reason we need these extended tests.

**Comment by csdechant** [comment]
> @csdechant Is there a reason we cannot just a steady state mms test for each Kernel individually? Testing with increasing coupling over an entire RF cycle seems like it is overkill for testing each Kernel and a bit redundant. I don't really see a need to have these tests be a transient simulation if the Kernel is not a time derivative. If we guarantee that each kernel works as expected on its own we should be able to guarantee the expected behaviour of our system, since MOOSE takes care of using multiple kernels in one system.

These MMS test were to verify Zapdos for the uses of RF discharges, so the manufactured solution should represent a simple, yet similar, solution of a RF discharge. The reason for that is one should not assume that since two separate parts of codes work as intended, that the coupled version will work and shouldn't be tested. This is similar to the idea between unit vs integration testing. While MOOSE combines the residuals and handles the Jacobians of multiple kernels, that does not ensure that the coupled model will behave as expected (just an example, what if a developer coded a variable or material as non-AD when it should be AD? The non-couple test will show no error, as there is no coupling of variables, and MOOSE doesn't know it should be AD, MOOSE will just provide an incorrect Jacobian).

>  I would like to advocate for a simpler testing approach. A single steady state test MMS test for each kernel where each test uses only as many objects as strictly necessary for testing a specific kernel, ideally a single kernel per input, if possible. I would also like to see our testing convention be something closer to the MOOSE convection of separating the tests by what type of object we are testing (Kernel, AuxKernel, ..., etc.).

I agree and disagree with this to a point (maybe it is just wording). My view is similar, as Zapdos should have more simple cases (similar to the MOOSE framework), but also my rigid coupled cases (similar to other MOOSE modules, such as the Navier Stokes module). My vision for the Zapdos testbed is as such:
- Have a type of "unit test" system that test just one kernel at a time (This is not a "true" unit test, as these are not testing single functions). These would be strictly non-heavy tests. 
-  Have a type of integration test system that test the coupling of multiple kernels and actions to verify the intended models of Zapdos. This would be a mixture of non-heavy and heavy tests, depending on run time. These should be coarse version of more rigid verification tests.
- Have a system of verification tests. These are full convergence study test that compare the CSV files output and plot convergence slopes. These are marked as heavy, by default, and will generate verification/convergences plots for the Zapdos website (similar to TMAP8).
- Finally, have a system of validation tests. Similar to the verifications test, these test will be marked as heavy, by default, and if need will be marked as hpc-heavy. These tests too will generate plots for the Zapdos website (similar to TMAP8 validation tests).

> The other concern, along with the size of gold files, I have with these tests is the amount of time it takes to run the test suite. Checking out the CI, testing, https://civet.inl.gov/job/2565410/, it shows that this test suite takes a little over 7 minutes to run. It also shows an average test time of 16 seconds, even when running the tests in parallel. I really think we should be aiming to decrease this to somewhere around a few seconds for each test, unless there is a specific reason we need these extended tests.

I just push a commit to reduce the gold files and run time. Also, the new MMS test are marked as heavy, so they will only run if a user inputs `./run_tests -jn --heavy`. It seems CIVET for Zapdos runs all the test together instead as a separate `Test Heavy`, like for MOOSE and MOOSE module PRs. @cticenhour Is there a setting to separate regular vs heavy test, so it doesn't seem like develops are submitting longer test without `heavy = True`?

@gsgall Please let me know if you have any question or different options as what I have stated above. @cticenhour We talked about this a little offline about revamping the Zapdos testbed, in regards to Issue #261. Please let my know if you have any comments or concerns about the about vision for the Zapdos testbed.

File: src/actions/AddPeriodicControllers.C [hunk 1]
```diff
@@ -138,77 +143,93 @@ AddPeriodicControllers::act()
...
-            (_start_time + _period) + _cycles_per_controls * _period * i + (_period * 0.0001);
+            (_start_time + _period * _cycles_per_controls) +
+            (_cycles_between_controls + _cycles_per_controls) * _period * i + (_period * 0.0001);
       }
 
       // If control cycles are back to back, then disable_start and disable_end times would be the
       // same, To avoid this, if cycles are back to back, then the disable_end starts at +
       // 0.0001*period
-      if (_cycles_per_controls == 1.0)
+      if (_cycles_between_controls == 1.0)
```
Comment (gsgall) [inline src/actions/AddPeriodicControllers.C:211]: We really shouldn't be doing this sort of float equality comparison. we should be using the absolute fuzzy equals functionality for this sort of comparison.

File: src/materials/FieldSolverMaterial.C [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  {
+    mooseError("Both electric field and potential variables have been provided. Please provide "
+               "only one, consistent with the solver mode parameter.");
+  }
+  else if (!parameters.isParamSetByUser("electric_field") &&
+           !parameters.isParamSetByUser("potential"))
+  {
+    mooseError("Neither potential nor electric field variables are defined. Please provide a "
+               "variable for field determination.");
+  }
```
Comment (gsgall) [inline src/materials/FieldSolverMaterial.C:63]: These should be using paramError instead of mooseErrors.

File: src/materials/Water.C [hunk 3]
```diff
@@ -330,7 +335,8 @@ Water::computeQpProperties()
 
   // Material version of aux variables
 
-  _EField[_qp] = -_grad_potential[_qp](0);
+  _EField[_qp] = _electric_field[_qp](0); // TODO: does this need to be 0 component now? consider
+                                          // fixing this up....1D will request the 0 by default
```
Comment (gsgall) [inline src/materials/Water.C:339]: are these going to be addressed in this PR or a later one?

**Review by gsgall — CHANGES_REQUESTED** [review]
In general most of my comments are small changes. 

There are a few things that I need to be address but this PR may not be the right place for them. But, here is a summary of those changes: 

- [ ] There are a lot of files in this PR which do not adhere to const correctness.
- [ ] Some files have MooseVariables as class member variables but only use those during construction. Unless there is a good reason those should not be member variables but temporary constructor variables for the class. 
- [ ] A lot of header files are missing proper documentation for all variables, functions and function parameters. 
- [ ] Some of the files which add new functionality do not fully support multi-ion systems. Nor do they use materials for secondary emission coefficients like the rest of the current Zapdos implementation.

**Comment by csdechant** [comment]
@gsgall I am working on this PR this week, but I did want to address one of your comments right now:

> * A lot of header files are missing proper documentation for all variables, functions and function parameters

Are you referring to any of the newly added header files or the previous header files? If it is the latter, that falls under PR https://github.com/shannon-lab/zapdos/pull/263 (which is ready for re-review). Depending on which PR is merged first, the other PR will most likely need to be rebased due to replacing `_grad_potential` with `_electric_field`. If possible, I would like PR https://github.com/shannon-lab/zapdos/pull/263 to be merged first.

File: src/bcs/SakiyamaEnergySecondaryElectronWithEffEfieldBC.C [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  _sgnip.resize(_num_ions);
+
+  for (unsigned int i = 0; i < _num_ions; ++i)
+  {
+    _ip_var[i] = getVar("ip", i);
+    _ip[i] = &adCoupledValue("ip", i);
+    _muip[i] = &getADMaterialProperty<Real>("mu" + (*getVar("ip", i)).name());
+    _sgnip[i] = &getMaterialProperty<Real>("sgn" + (*getVar("ip", i)).name());
+  }
+}
```
[the remark this answers is not in the document]
Comment (csdechant) [reply src/bcs/SakiyamaEnergySecondaryElectronWithEffEfieldBC.C:69]: After a quick look at the other BC that account of multi-species ions, quite of few of them have those MooseVariables in the Devel branch due to the pre-AD days. Because of this, and in the spirit of having smaller but more focus PRs, I suggest we make this an issue and I can address this in a separate PR.

File: src/bcs/SakiyamaSecondaryElectronWithEffEfieldBC.C [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+SakiyamaSecondaryElectronWithEffEfieldBC::validParams()
+{
+  InputParameters params = ADIntegratedBC::validParams();
+  params.addRequiredCoupledVar("ip", "The ion density.");
+  params.addRequiredCoupledVar("Ex", "The EField in the x-direction");
+  params.addCoupledVar("Ey", 0, "The EField in the y-direction"); // only required in 2D and 3D
+  params.addCoupledVar("Ez", 0, "The EField in the z-direction"); // only required in 3D
+  params.addRequiredParam<Real>("position_units", "Units of position.");
+  params.addParam<Real>("users_gamma",
+                        "A secondary electron emission coeff. only used for this BC.");
```
[the remark this answers is not in the document]
Comment (csdechant) [reply src/bcs/SakiyamaSecondaryElectronWithEffEfieldBC.C:25]: Same as the comment response to `DielectricBCWithEffEfield.C` and will address in this PR.

File: src/materials/Water.C (hunk 3 above)
Comment (csdechant) [reply src/materials/Water.C:339]: This object was made during Alex's thesis and I haven't used it except when I was first using Zapdos. After a quick search in Zapdos, it doesn't seem that any of the test files used this exact material property. That being said, I don't think we should remove it for this PR, as this and other material objects (such as `Gas.C` and `GasElectronMoments`) should be re-examined to see if any declare property are not being utilized in a current test and address the need for a test or the removal of the property.

**Review by csdechant — COMMENTED** [review]
@gsgall I am working on addressing your comments, but I have addressed some of your questions regarding if a change should be made in this PR or a separate PR (as I believe some of the comment should be address in a separate PR, in the spirit of small but more focused PR in Zapdos' future). Please let me know if you have a difference opinion on which issues belong in this PR or a separate PR.

**Comment by gsgall** [comment]
> @gsgall I am working on addressing your comments, but I have addressed some of your questions regarding if a change should be made in this PR or a separate PR (as I believe some of the comment should be address in a separate PR, in the spirit of small but more focused PR in Zapdos' future). Please let me know if you have a difference opinion on which issues belong in this PR or a separate PR.

Thank you @csdechant. I think your comments and wanting to make these changes in a separate PR a reasonable. I don't have any issue with this plan.

File: src/postprocessors/SideCurrent.C [hunk 6]
```diff
@@ -87,15 +92,15 @@ SideCurrent::computeQpIntegral()
...
-    _ion_flux += 0.5 *
-                     std::sqrt(8 * _kb[_qp] * (*_T_ions[i])[_qp] / (M_PI * (*_mass_ions[i])[_qp])) *
-                     std::exp((*_ions[i])[_qp]) +
-                 (2 * _a - 1) * (*_sgn_ions[i])[_qp] * (*_mu_ions[i])[_qp] * -_grad_potential[_qp] *
-                     _r_units * std::exp((*_ions[i])[_qp]) * _normals[_qp];
+    _ion_flux +=
+        0.5 * std::sqrt(8 * _kb[_qp] * (*_T_ions[i])[_qp] / (M_PI * (*_mass_ions[i])[_qp])) *
+            std::exp((*_ions[i])[_qp]) +
+        (2 * _a - 1) * (*_sgn_ions[i])[_qp] * (*_mu_ions[i])[_qp] *
+            raw_value(_electric_field[_qp]) * _r_units * std::exp((*_ions[i])[_qp]) * _normals[_qp];
```
Comment (gsgall) [inline src/postprocessors/SideCurrent.C:99]: Just curious with this one but why do we need the raw values here?

**Review by gsgall — CHANGES_REQUESTED** [review]
Just a few really small things. 

@csdechant @cticenhour The only thing I am unsure about is wether or not we want to enforce the proper use of the `requirement` field within the test files. So whichever you guys prefer is fine with me on this one.

File: test/tests/field_solver/tests [hunk 7]
```diff
@@ -0,0 +1,41 @@
+[Tests]
+  [./electrostatic]
+    type = Exodiff
+    input = 'field_solver_material_electrostatic.i'
+    exodiff = 'field_solver_material_electrostatic_out.e'
+    custom_cmp = 'field_solver_material_electrostatic_out.cmp'
+  [../]
```
[the remark this answers is not in the document]
Comment (csdechant) [reply test/tests/field_solver/tests:7]: That is a great suggestion! I would like to hold off on that for now, since we lack full coverage on documentation, for now. I believe this should be added to issue #260.

[the remark this answers is not in the document]
Comment (csdechant) [reply test/tests/field_solver/tests:7]: Just added "Enforcing the requirement documentation on tests" to issue #260

**Review by gsgall — APPROVED** [review]
Thank you Corey!
